src/Microsoft.AspNetCore.Razor.Language/RazorSourceDocument.cs

```diff
@@ -183,13 +183,20 @@ public static RazorSourceDocument ReadFrom(RazorProjectItem projectItem)
             if (string.IsNullOrEmpty(filePath))
             {
                 // Fall back to the relative path only if necessary.
+                filePath = projectItem.RelativePhysicalPath;
+            }
+
+            if (string.IsNullOrEmpty(filePath))
+            {
+                // Then fall back to the FilePath (yeah it's a bad name) which is like an MVC view engine path
+                // It's much better to have something than nothing.
                 filePath = projectItem.FilePath;
             }
 
             using (var stream = projectItem.Read())
             {
                 // Autodetect the encoding.
-                return new StreamSourceDocument(stream, null, new RazorSourceDocumentProperties(filePath, projectItem.FilePath));
+                return new StreamSourceDocument(stream, null, new RazorSourceDocumentProperties(filePath, projectItem.RelativePhysicalPath));
```

Comment:
This was supposed to be updated in a previous commit but it got missed somehow.

Reply:
This was causing some codegen issues around recompilation, which is how I found it.

Reply:
The codegen tests don't use the template engine, and already had these documents created the right way.